File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/PublicKeyCredentialSource.java
/*
 * Copyright (c) 2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth.webauthn;

import android.os.Parcel;
import android.os.Parcelable;
import android.util.Base64;

import com.google.android.gms.fido.fido2.api.common.PublicKeyCredentialDescriptor;
import com.google.android.gms.fido.fido2.api.common.PublicKeyCredentialType;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Arrays;
import java.util.Collections;

import lombok.Builder;
import lombok.Getter;

import static com.google.android.gms.fido.common.Transport.INTERNAL;

/**
 * Representation of Public Key Credential Source that received after WebAuthn registration.
 */
@Getter
public class PublicKeyCredentialSource implements Parcelable {

    //Credential ID
    private final byte[] id;
    private String type;
    private final String rpid;
    private final byte[] userHandle;
    private final String otherUI;

    @Builder
    public PublicKeyCredentialSource(byte[] id, String type, String rpid, byte[] userHandle, String otherUI) {
        this.id = id;
        this.type = type;
        if (type == null) {
            this.type = PublicKeyCredentialType.PUBLIC_KEY.toString();
        }
        this.rpid = rpid;
        this.userHandle = userHandle;
        this.otherUI = otherUI;
    }

    protected PublicKeyCredentialSource(Parcel in) {
        id = in.createByteArray();
        type = in.readString();
        rpid = in.readString();
        userHandle = in.createByteArray();
        otherUI = in.readString();
    }

    public boolean equalsUserHandle(PublicKeyCredentialSource source) {
        return Arrays.equals(userHandle, source.getUserHandle());
    }

    public static final Creator<PublicKeyCredentialSource> CREATOR = new Creator<PublicKeyCredentialSource>() {
        @Override
        public PublicKeyCredentialSource createFromParcel(Parcel in) {
            return new PublicKeyCredentialSource(in);
        }

        @Override
        public PublicKeyCredentialSource[] newArray(int size) {
            return new PublicKeyCredentialSource[size];
        }
    };

    /**
     * Convert {@link PublicKeyCredentialSource} to {@link JSONObject}
     *
     * @return The result JSONObject
     */
    public JSONObject toJson() {
        JSONObject result = new JSONObject();
        try {
            result.put("id", Base64.encodeToString(id, Base64.DEFAULT));
            result.put("type", type);
            result.put("rpid", rpid);
            if (userHandle != null) {
                result.put("userHandle", Base64.encodeToString(userHandle, Base64.DEFAULT));
            }
            result.put("otherUI", otherUI);
            return result;
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Convert {@link PublicKeyCredentialSource} to {@link PublicKeyCredentialDescriptor}
     *
     * @return The result {@link PublicKeyCredentialDescriptor}
     */
    public PublicKeyCredentialDescriptor toDescriptor() {
        try {
            return new PublicKeyCredentialDescriptor(
                    PublicKeyCredentialType.fromString(type).toString(),
                    id,
                    Collections.singletonList(INTERNAL));
        } catch (PublicKeyCredentialType.UnsupportedPublicKeyCredTypeException e) {
            throw new UnsupportedOperationException(e);
        }
    }

    /**
     * Convert {@link JSONObject} to {@link PublicKeyCredentialSource}
     *
     * @param jsonObject The json object to be converted
     * @return The result PublicKeyCredentialSource
     */
    public static PublicKeyCredentialSource fromJson(JSONObject jsonObject) {
        try {
            return PublicKeyCredentialSource.builder()
                    .id(Base64.decode(jsonObject.getString("id"), Base64.DEFAULT))
                    .type(jsonObject.optString("type", null))
                    .rpid(jsonObject.optString("rpid", null))
                    .userHandle(jsonObject.has("userHandle") ? Base64.decode(jsonObject.getString("userHandle"), Base64.DEFAULT) : null)
                    .otherUI(jsonObject.optString("otherUI", null))
                    .build();
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeByteArray(id);
        dest.writeString(type);
        dest.writeString(rpid);
        dest.writeByteArray(userHandle);
        dest.writeString(otherUI);
    }
}

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/PublicKeyCredentialSource.kt
/*
 * Copyright (c) 2022 ForgeRock. All rights reserved.
 *
 *  This software may be modified and distributed under the terms
 *  of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth.webauthn

import android.os.Parcelable
import android.util.Base64
import com.google.android.gms.fido.common.Transport
import com.google.android.gms.fido.fido2.api.common.PublicKeyCredentialDescriptor
import com.google.android.gms.fido.fido2.api.common.PublicKeyCredentialType
import kotlinx.parcelize.Parcelize
import org.json.JSONObject

private const val ID = "id"

private const val TYPE = "type"

private const val RPID = "rpid"

private const val USER_HANDLE = "userHandle"

private const val OTHER_UI = "otherUI"

private const val CREATED = "created"

/**
 * Representation of Public Key Credential Source that received after WebAuthn registration.
 */
@Parcelize
class PublicKeyCredentialSource(val id: ByteArray,
                                val type: String = PublicKeyCredentialType.PUBLIC_KEY.toString(),
                                val rpid: String,
                                val userHandle: ByteArray,
                                val otherUI: String,
                                val created: Long) : Parcelable {

    /**
     * Convert [PublicKeyCredentialSource] to [JSONObject]
     *
     * @return The result JSONObject
     */
    fun toJson() = JSONObject().apply {
        this.put(ID, Base64.encodeToString(id, Base64.DEFAULT))
        this.put(TYPE, type)
        this.put(RPID, rpid)
        this.put(USER_HANDLE, Base64.encodeToString(userHandle, Base64.DEFAULT))
        this.put(OTHER_UI, otherUI)
        this.put(CREATED, created)
    }

    /**
     * Convert [PublicKeyCredentialSource] to [PublicKeyCredentialDescriptor]
     *
     * @return The result [PublicKeyCredentialDescriptor]
     */
    fun toDescriptor() =
        PublicKeyCredentialDescriptor(PublicKeyCredentialType.fromString(type).toString(),
            id, listOf(Transport.INTERNAL))

    class PublicKeyCredentialSourceBuilder internal constructor() {
        private lateinit var id: ByteArray
        private var type: String = PublicKeyCredentialType.PUBLIC_KEY.toString()
        private lateinit var rpid: String
        private lateinit var userHandle: ByteArray
        private lateinit var otherUI: String
        private var created: Long? = null

        fun id(id: ByteArray): PublicKeyCredentialSourceBuilder {
            this.id = id
            return this
        }

        fun type(type: String): PublicKeyCredentialSourceBuilder {
            this.type = type
            return this
        }

        fun rpid(rpid: String): PublicKeyCredentialSourceBuilder {
            this.rpid = rpid
            return this
        }

        fun userHandle(userHandle: ByteArray): PublicKeyCredentialSourceBuilder {
            this.userHandle = userHandle
            return this
        }

        fun otherUI(otherUI: String): PublicKeyCredentialSourceBuilder {
            this.otherUI = otherUI
            return this
        }

        fun created(created: Long): PublicKeyCredentialSourceBuilder {
            this.created = created
            return this
        }

        fun build(): PublicKeyCredentialSource {
            return PublicKeyCredentialSource(id, type, rpid, userHandle, otherUI, created = created ?: System.currentTimeMillis())
        }
    }

    companion object {
        fun builder() = PublicKeyCredentialSourceBuilder()

        /**
         * Convert [JSONObject] to [PublicKeyCredentialSource]
         *
         * @param jsonObject The json object to be converted
         * @return The result PublicKeyCredentialSource
         */
        @JvmStatic
        fun fromJson(jsonObject: JSONObject) =
            builder()
                .id(Base64.decode(jsonObject.getString(ID), Base64.DEFAULT))
                .type(jsonObject.getString(TYPE))
                .rpid(jsonObject.getString(RPID))
                .userHandle(Base64.decode(jsonObject.getString(USER_HANDLE), Base64.DEFAULT))
                .otherUI(jsonObject.getString(OTHER_UI))
                .created(jsonObject.optLong(CREATED,0 ))
                .build()

    }
}